ores/lock_manager/poolcounter.py

```diff
@@ -0,0 +1,58 @@
+"""
+Class to implement PoolCounter lock manager.
+
+This is useful to make sure too many connections are not coming from one IP.
+"""
+
+import socket
+
+from .lock_manager import LockManager
+
+
+class PoolCounter(LockManager):
+    def __init__(self, config, logger):
+        self.server = config['server']
+        self.port = config['port']
+        self.stream = None
+        self.logger = logger
+
+    def connect(self):
+        self.stream = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+        return self.stream.connect((self.server, self.port))
+
+    def acq4me(self, key, workers, maxqueue, timeout):
+        if not self.stream:
+            self.connect()
+        try:
+            self.stream.send(bytes('ACQ4ME %s %d %d %d\n' % (key, workers, maxqueue, timeout), 'utf-8'))
+            data = self.stream.recv(4096).decode('utf-8')
+            print(data)
+        except socket.error as e:
+            self.stream = None
+            raise e
+
+        return data == 'LOCKED\n'
+
+    def release(self, key):
+        if not self.stream:
```

Comment:
Yes, but this sanity check makes sure we don't try to run "send" method on None object which causes the whole thing to fall apart. It's better to be there IMO (Also it's copied from Gilles' version on Django: https://github.com/wikimedia/operations-debs-python-thumbor-wikimedia/blob/master/wikimedia_thumbor/poolcounter/__init__.py )

Reply:
I haven't interacted with PoolCounter personally, but the wikitech [page](https://wikitech.wikimedia.org/wiki/PoolCounter) says,

> The network protocol is line-based, with parameters separated by spaces (spaces in parameters are percent-encoded). The client opens a connection, sends a lock acquire command, does the work, sends a lock release command, then closes the connection. 

Maybe the docs are wrong?

Reply:
Sorry, I see what you're saying now.  If the connection was dropped, then the docs are ambiguous about whether we need to reconnect and release, or if the server will release automatically.

Reply:
This page is trying to be the client so the responsibility of opening the connection, sending things, and closing it is on this class.

Reply:
The server uses open connections as lock identifiers. I don't think it has any way to correlate two connections coming from the same client. So reconnecting and releasing (disconnecting) is just a costly noop, you could just return true instead.

Although arguably you should throw an exception since it signifies some kind of logic error in the code (trying to release a lock that was never obtained).

Reply:
> Although arguably you should throw an exception since it signifies some kind of logic error in the code (trying to release a lock that was never obtained).

I like this idea, but @Ladsgroup is trying to *not* blow up because higher-level code should be the judge of whether its response will be valid, I think.